packages/medusa/src/repositories/discount-condition.ts

```diff
@@ -108,39 +116,166 @@ export class DiscountConditionRepository extends Repository<DiscountCondition> {
+    )[]
+  > {
+    const { joinTableName, joinTableForeignKey } =
+      this.getJoinTableResourceIdentifiers(type)
+
+    if (joinTableName) {
+      return await this.manager
+        .createQueryBuilder(joinTableName, "conds")
+        .select()
+        .where(`${"conds"}.${joinTableForeignKey} = :resourceId`, {
+          resourceId,
+        })
+        .andWhere(`${"conds"}.condition_id = :condId`, {
+          condId,
+        })
+        .getMany()
+    }
+
+    return []
+  }
+
+  async isValidForProduct(
+    discountRuleId: string,
+    product: Product
+  ): Promise<boolean> {
+    const discountConditions = await this.createQueryBuilder("discon")
+      .select(["discon.id", "discon.type", "discon.operator"])
+      .where("discon.discount_rule_id = :discountRuleId", {
+        discountRuleId,
+      })
+      .getMany()
+
+    // in case of no discount conditions, we assume that the discount
+    // is valid for all
+    if (!discountConditions.length) {
+      return true
+    }
+
+    // retrieve all conditions for each type where condition type id is in jointable (products, product_types, product_collections, product_tags)
+    // "E.g. for a given product condition, give me all products affected by it"
+    // for each of these types, we check:
+    //    if condition operation is `in` and the query for conditions defined for the given type is empty, the discount is invalid
+    //    if condition operation is `not_in` and the query for conditions defined for the given type is not empty, the discount is invalid
+    for (const condition of discountConditions) {
+      const { joinTableForeignKey } = this.getJoinTableResourceIdentifiers(
+        condition.type
+      )
+
+      // Tags is a ManyToMany relation, so we need to check all tags that exist
+      // For each tag, we perform the operation outlined in above comment
+      if (
+        joinTableForeignKey ===
+        DiscountConditionJoinTableForeignKey.PRODUCT_TAG_ID
+      ) {
+        for (const tag of product.tags) {
+          const tagConds = await this.queryConditionTable({
+            type: condition.type,
+            condId: condition.id,
+            resourceId: tag.id,
+          })
+
+          if (
+            condition.operator === DiscountConditionOperator.IN &&
+            !tagConds.length
+          ) {
+            return false
+          }
+
+          if (
+            condition.operator === DiscountConditionOperator.NOT_IN &&
+            tagConds.length
+          ) {
+            return false
+          }
+        }
+      } else {
+        let resourceId: string | null = product.id
+
+        if (condition.type === DiscountConditionType.PRODUCT_TYPES) {
+          resourceId = product.type_id
+        }
+
+        if (condition.type === DiscountConditionType.PRODUCT_COLLECTIONS) {
+          resourceId = product.collection_id
+        }
+
+        if (!resourceId) {
+          return false
+        }
```

Comment:
Added this snippet @zakariaelas. The condition can only have one type of resource, but we still need to query for the specific resource id, which I did not include in the previous PR. 

So if `condition.type === "products"`, we need to use `product.id` to query for. Does that make sense?